Add query index management support

alfinkel opened the pull request:
_What:_

Provide API to manage Cloudant Query JSON and TEXT indexes.  This PR contains only the code/logic for this functionality.  A follow-up PR(s) will include the unit tests which are complete and all run to successful completion.  The work has been split up to make the work easier to review.

_Why:_

In order to successfully use Cloudant Query, the ability to create and delete query indexes needs to be provided.

_How:_

- Add an indexes module that handles the creation and deletion of Cloudant Query indexes.
- Modify the views module and add a QueryIndexView class to handle the representation of a JSON index as a view in a design document.
- Modify the design_documents module to handle the new views module updates.
- Add API methods to the database module that provide an interface to creating an index, deleting an index and retrieving the list of all indexes in the database.

_Issues:_

- This PR specifically addresses #40.
- As a side effect, #38 is also addressed as part of this PR.

reviewer: @evansde77 
reviewer: @emlaver

evansde77 commented on src/cloudant/database.py:
```diff
@@ -792,6 +793,128 @@ def shards(self):
+            return resp.json()
+        
+        indexes = []
+        for data in resp.json().get('indexes', None):
```
since you are going to iterate over this, you may want [] instead of None, since None isnt iterable

alfinkel replied:
Fixed in 5740aa251ca744ad62d53bb36986bf68e2da4340.

evansde77 commented on src/cloudant/database.py:
```diff
@@ -792,6 +793,128 @@ def shards(self):
+            remote database
+        """
+        index = None
+        if index_type == 'json':
```
I would suggest defining a set of index type module constants like JSON_INDEX_TYPE='json' and using those constants in the code.

alfinkel replied:
Added an index_constants module in 21c9d5dd5d95925c23a67fe9318f26c8a10e07b3 and changed all references in the database module to reference the index constants in 5740aa251ca744ad62d53bb36986bf68e2da4340.

evansde77 commented on src/cloudant/design_document.py:
```diff
@@ -86,14 +90,15 @@ def update_view(self, view_name, map_func, reduce_func=None):
         Removes a View from the locally cached DesignDocument View dictionary.
+        Do not use this method if you wish to
+        delete a query index view.  Use
+        :func:`~cloudant.database.CloudantDatabase.delete_index` instead.
```
what happens if they try it any way? Execption? Maybe call that out in the doc. Otherwise this is just a "do not push this button" type of button

alfinkel replied:
I actually added some more validation to `add_view`, `update_view`, and `delete_view` that raises an exception if a QueryIndexView is found and added more detail to the docstrings as well.  Done in 2707ad460db815447a917e39d8b0d67536a7ea3f.

evansde77 commented on src/cloudant/design_document.py:
```diff
@@ -102,19 +107,53 @@ def delete_view(self, view_name):
-                view_def.get('map'),
-                view_def.get('reduce')
-            )
+            if self.get('language', None) != 'query':
```
would make query a module level constant

alfinkel replied:
Done in 2707ad460db815447a917e39d8b0d67536a7ea3f.

evansde77 commented on src/cloudant/indexes.py:
```diff
@@ -0,0 +1,271 @@
+    	"""
+    	Creates the current index in the remote database.
+    	"""
+        payload = {'type': self._type}
```
indentation is off on the docstring

alfinkel replied:
Fixed in daf19a6dd600acb796fb87dc7f78a0431b5940dc.

evansde77 commented on src/cloudant/indexes.py:
```diff
@@ -0,0 +1,271 @@
+                'contains the \"fields\" argument.'
+            ).format(kwargs)
+            raise CloudantArgumentError(msg)
+        if kwargs['fields'] != [{'_id': 'asc'}]:
```
This is a pretty specific kwarg, since it is limited, I would maybe make this a boolean in the call like ascending_order=True and then if it is True, just set the fields content to {'_id': 'asc'} 
Make the API feel more like an API than exposing deeper knowledge of the special index, goal is to make it easy for the library user.

alfinkel replied:
After reading your comment and giving some more thought to the actual purpose of this class, I decided to refactor it slightly.  Essentially the purpose of a SpecialIndex in this library is for use by the database API method `get_all_indexes()`.  'get_all_indexes()` retrieves the indexes in the database through the _index endpoint and wraps the primary "special" index as a SpecialIndex object as well as wrapping JSON indexes as Index objects and TEXT indexes as SearchIndex objects and returns the index list to the caller.

Since you cannot really do anything with a SpecialIndex object other than instanciating it, it didn't really make much sense to have any kind of validation as was originally being done.  So the validation was removed and the method signature was changed to match that if the parent Index class.  This change provides us with flexibility if we (Cloudant) decide to change the "special" index definition server side at any point down the road (unlikely but still possible), this library implementation will still be able to handle it.

This is done in daf19a6dd600acb796fb87dc7f78a0431b5940dc and 5740aa251ca744ad62d53bb36986bf68e2da4340.

evansde77 commented on src/cloudant/indexes.py:
```diff
@@ -0,0 +1,271 @@
+            '_all_docs',
+            **kwargs
+        )
+        self._type = 'special'
```
more constant style stuff. Looks like having an index_constants.py module with these all defined would make it a bit cleaner & easier to determine actual constants from magic strings in the code base.

alfinkel replied:
Done in 21c9d5dd5d95925c23a67fe9318f26c8a10e07b3 and daf19a6dd600acb796fb87dc7f78a0431b5940dc.

evansde77 commented on src/cloudant/views.py:
```diff
@@ -325,3 +334,94 @@ def custom_result(self, **options):
+            'use the database \'get_query_result\' convenience method.'
+        )
+
+    def make_result(self, **options):
```
Guess this is overriding a base method that doesnt make sense for this object, would maybe spell that out in the docstring.

alfinkel replied:
Added more detail to the docstring for this method.  Done in b235b950d4efc80ddce6d75cfdf0ee93e48126f7.

evansde77 commented:
Mostly stylistic/documentation comments. Hopefully unittests are coming as well?

alfinkel commented:
Unit tests and doc updates are on the way in follow up PR(s).  The work is complete and the unit tests all ran successfully.  I split the work up into separate PRs to hopefully make the review process easier.  I updated the PR description to include this info.

evansde77 commented on src/cloudant/indexes.py:
```diff
@@ -0,0 +1,260 @@
+    	Constructs and returns the index URL.
+
+    	:returns: Index URL
+    	"""
```
indent

evansde77 commented:
LGTM, although there are a couple of indent/spacing things, might be worth running pep8 over it, see what it tells you.

alfinkel commented:
I ran pep8 and all should be good with indentation and spacing now.  It looks like I a mixture of tabs and spaces in Sublime Text was the cause.  Obviously I need to fix my indenting configuration in the editor.  Thanks for catching that.

emlaver commented:
+1
